packages/atlas-signin/src/main.ts

```diff
@@ -0,0 +1,142 @@
+import { ipcMain, shell } from 'electron';
+import { URLSearchParams } from 'url';
+import * as plugin from '@mongodb-js/oidc-plugin';
+import fetch from 'node-fetch';
+import type { IntrospectInfo, Token, UserInfo } from './shared';
+import { Events } from './shared';
+
+export class AtlasSignIn {
+  private constructor() {
+    // singleton
+  }
+
+  private static calledOnce = false;
+
+  private static plugin = plugin.createMongoDBOIDCPlugin({
+    // TODO: as auth is done through the localhost loopback and not directly
+    // through customer okta, we will need to implement screens for following
+    // states: initial, success, error, not found; waiting for the designs
+    // redirectServerRequestHandler() {
+    // }
+    async openBrowser({ url }) {
+      await shell.openExternal(url);
+    },
+  });
+
+  private static token: Token | null = null;
+
+  private static get clientId() {
+    if (!process.env.COMPASS_CLIENT_ID) {
+      throw new Error('COMPASS_CLIENT_ID is required');
+    }
+    return process.env.COMPASS_CLIENT_ID;
+  }
+
+  private static get issuer() {
+    if (!process.env.COMPASS_OIDC_ISSUER) {
+      throw new Error('COMPASS_OIDC_ISSUER is required');
+    }
+    return process.env.COMPASS_OIDC_ISSUER;
+  }
+
+  static init() {
+    if (this.calledOnce) {
+      return;
+    }
+    this.calledOnce = true;
+    ipcMain.handle(Events.SignIn, async () => {
+      this.token =
+        // TODO: provide better interface in the oidc-plugin so that we don't
+        // need to call internal callbacks
+        await this.plugin.mongoClientOptions.authMechanismProperties.REQUEST_TOKEN_CALLBACK(
+          { clientId: this.clientId, issuer: this.issuer },
+          // Required driver specific stuff
+          { version: 0 }
+        );
+      this.setupTokenRefresh();
+      return this.token;
+    });
+    ipcMain.handle(Events.IsAuthenticated, async () => {
+      if (!this.token) {
+        return false;
+      }
+      try {
+        return (await this.introspect()).active;
+      } catch (err) {
+        return false;
+      }
+    });
+    ipcMain.handle(Events.UserInfo, async () => {
+      return await this.getUserInfo();
+    });
+  }
+
+  // TODO: good-ish enough for POC, but we will have reimplement something
+  // similar to https://github.com/okta/okta-auth-js#tokenmanageronevent-callback-context
+  private static setupTokenRefresh() {
+    setTimeout(
+      () => {
+        void this.plugin.mongoClientOptions.authMechanismProperties
+          .REFRESH_TOKEN_CALLBACK(
+            { clientId: this.clientId, issuer: this.issuer },
+            // Required driver specific stuff
+            { version: 0 }
+          )
+          .then((token) => {
+            this.token = token;
+            this.setupTokenRefresh();
+          });
+      },
+      Math.max(
+        // Min idle is 10mins
+        10 * 60 * 1000,
+        ((this.token?.expiresInSeconds ?? 0) * 1000) / 2
+      )
+    ).unref();
```

Comment:
The oidc-plugin code already implements automatic refresh, I don’t think you need this function:

https://github.com/mongodb-js/oidc-plugin/blob/48ddba454c49a84b50a7b27b7513f6ce274109df/src/plugin.ts#L459-L470

Reply:
Cool, good to know